tests/devices/qubit/test_jaxpr_adjoint_jvp.py

```diff
@@ -0,0 +1,246 @@
+
+        def f(x):
+            qml.RX(x, 0)
+            return qml.expval(qml.Z(0))
+
+        jaxpr = jax.make_jaxpr(f)(0.5)
+
+        args = (0.82,)
+        tangents = (2.0,)
+
+        executor = partial(execute_and_jvp, jaxpr.jaxpr, num_wires=1)
+        if use_jit:
+            executor = jax.jit(executor)
+
+        results, dresults = executor(args, tangents)
+
+        assert len(results) == 1
+        assert qml.math.allclose(results, jnp.cos(args[0]))
+        assert len(dresults) == 1
+        assert qml.math.allclose(dresults[0], tangents[0] * -jnp.sin(args[0]))
+
+    def test_multiple_in(self):
+        """Test that we can differentiaite multiple inputs."""
+
+        def f(x, y):
+            qml.RX(x, 0)
+            qml.RY(y, 1)
+            qml.CNOT((0, 1))
+            return qml.expval(qml.Y(0))
+
+        x = jnp.array(0.5)
+        y = jnp.array(1.2)
+        dx = jnp.array(2.0)
+        dy = jnp.array(3.0)
+
+        jaxpr = jax.make_jaxpr(f)(x, y).jaxpr
+
+        [res], [dres] = execute_and_jvp(jaxpr, (x, y), (dx, dy), num_wires=2)
+
+        expected = -jnp.sin(x) * jnp.sin(y)
+        assert qml.math.allclose(res, expected)
+
+        expected_dres = dx * -jnp.cos(x) * jnp.sin(y) + dy * -jnp.sin(x) * jnp.cos(y)
+        assert qml.math.allclose(dres, expected_dres)
+
+    def test_multiple_output(self):
+        """Test we can compute the jvp with multiple outputs."""
+
+        def f(x):
+            qml.RX(x, 0)
+            return qml.expval(qml.X(0)), qml.expval(qml.Y(0)), qml.expval(qml.Z(0))
+
+        jaxpr = jax.make_jaxpr(f)(0.5).jaxpr
+
+        x = -0.5
+        res, dres = execute_and_jvp(jaxpr, (x,), (2.0,), num_wires=1)
+
+        assert qml.math.allclose(res[0], 0)
+        assert qml.math.allclose(res[1], -jnp.sin(x))
+        assert qml.math.allclose(res[2], jnp.cos(x))
+
+        assert qml.math.allclose(dres[0], 0)
+        assert qml.math.allclose(dres[1], 2.0 * -jnp.cos(x))
+        assert qml.math.allclose(dres[2], 2.0 * -jnp.sin(x))
+
+    def test_classical_preprocessing(self):
+        """Test that we can perform classical preprocessing of variables."""
+
+        def f(x):
+            y = x**2
+            qml.RX(y[0], 0)
+            qml.RX(y[1], 1)
+            return qml.expval(qml.Z(0) @ qml.Z(1))
+
+        x = jnp.array([1.5, 2.5])
+        dx = jnp.array([2.0, 3.0])
+        jaxpr = jax.make_jaxpr(f)(x).jaxpr
+
+        [res], [dres] = execute_and_jvp(jaxpr, (x,), (dx,), num_wires=2)
+
+        expected = jnp.cos(x[0] ** 2) * jnp.cos(x[1] ** 2)
+        assert qml.math.allclose(res, expected)
+        dexpected = (
+            -jnp.sin(x[0] ** 2) * 2 * x[0] * dx[0] * jnp.cos(x[1] ** 2)
+            + jnp.cos(x[0] ** 2) * -jnp.sin(x[1] ** 2) * 2 * x[1] * dx[1]
+        )
+        assert qml.math.allclose(dres, dexpected)
+
+    def test_jaxpr_consts(self):
+        """Test that we can execute jaxpr with consts."""
```

Comment:
Are closed-over variables supported already as well? :) Could test that in the same test, possibly

Reply:
consts and closed over variables are essentially the same thing from the jaxpr perspective. They both go into `jaxpr.consts`. Since  I'm testing executing the jaxpr, and not capturing it, I think one way of producing `consts` should be fine.